### PR TITLE
add new hlsl cpp codegen

### PR DESCRIPTION
**Pull request by xiayuqing0622**
1.add new hlsl cpp codegen （use `-fhlsl_codegen_type=cpp`)
`-fhlsl_codegen_type` chooses from [default(the old hlsl codegen, will be deprecated, csharp, cpp] .` -fhlsl_codegen_type=csharp` is equivalent to `-fhlsl_csharp_codegen=true`, and` -fhlsl_csharp_codegen` is deprecated in this pr.
2. change `LanguageUnitwithVec::unit_vec` type from vector to deque
3. add customized memory implementation option. use `-fcustomized_mem_imp=true` to turn on.
4. eliminate code check for files under template folder

### CODE REVIEW

**Review by wenxcs — DISMISSED** [review]

File: src/tools/nnfusion/templates/dxcompute/Direct3DWinNN/Direct3DWinNN.sln [hunk 1]
```diff
@@ -0,0 +1,31 @@
+﻿
```
Comment (jlxue) [inline src/tools/nnfusion/templates/dxcompute/Direct3DWinNN/Direct3DWinNN.sln:1]: Seems we do not need to include .sln, vcxproj.user, vcxproj.filters files

File: src/tools/nnfusion/templates/dxcompute/Direct3DXBoxNN/DeviceResources.cpp [hunk 2]
```diff
@@ -0,0 +1,444 @@
+//
+// DeviceResources.cpp - A wrapper for the Direct3D 12.X device and swapchain
+//
+
```
Comment (jlxue) [inline src/tools/nnfusion/templates/dxcompute/Direct3DXBoxNN/DeviceResources.cpp:4]: Our code format change the order of this two headers, which causes a compiling error for xbox env. Can we just eliminate the code check under this template folder?

File: src/tools/nnfusion/templates/dxcompute/Direct3DWinNN/Direct3DWinNN.sln (hunk 1 above)
Comment (xiayuqing0622) [reply src/tools/nnfusion/templates/dxcompute/Direct3DWinNN/Direct3DWinNN.sln:1]: > Seems we do not need to include .sln, vcxproj.user, vcxproj.filters files

then we could remove these files from templates

File: src/tools/nnfusion/templates/dxcompute/Direct3DXBoxNN/DeviceResources.cpp (hunk 2 above)
Comment (xiayuqing0622) [reply src/tools/nnfusion/templates/dxcompute/Direct3DXBoxNN/DeviceResources.cpp:4]: sure

**Review by jlxue — APPROVED** [review]

**Pull request merged**